src/tpm2/crypto/openssl/CryptSym.c

```diff
@@ -458,6 +461,262 @@ CryptSymmetricDecrypt(
+    }
+
+#if ALG_TDES && ALG_CTR
+    if (algorithm == TPM_ALG_TDES && mode == ALG_CTR_VALUE) {
+        TDES_CTR(keyToUse, keyToUseLen * 8, dSize, dIn, iv, pOut, blockSize);
+        if (pOut != dOut)
+            memcpy(dOut, pOut, dSize);
+        ERROR_RETURN(TPM_RC_SUCCESS);
+    }
+#endif
+
+    ctx = EVP_CIPHER_CTX_new();
+    if (!ctx)
+        ERROR_RETURN(TPM_RC_FAILURE);
+
+    if (EVP_EncryptInit_ex(ctx, evpfn(), NULL, keyToUse, iv) != 1)
+        ERROR_RETURN(TPM_RC_FAILURE);
+
+    EVP_CIPHER_CTX_set_flags(ctx, EVP_CIPH_NO_PADDING);
+    if (EVP_EncryptUpdate(ctx, pOut, &outlen1, dIn, dSize) != 1)
+        ERROR_RETURN(TPM_RC_FAILURE);
+
+    EVP_CIPHER_CTX_set_flags(ctx, EVP_CIPH_NO_PADDING);
+    if (EVP_EncryptFinal_ex(ctx, pOut + outlen1, &outlen2) != 1)
+        ERROR_RETURN(TPM_RC_FAILURE);
+
+    pAssert(dSize >= outlen1 + outlen2);
+
+    if (pOut != dOut)
+        memcpy(dOut, pOut, outlen1 + outlen2);
+
+ Exit:
+    clear_and_free(buffer, buffersize);
+    EVP_CIPHER_CTX_free(ctx);
+
+    return retVal;
+}
+
+/* 10.2.20.5.1 CryptSymmetricDecrypt() */
+/* This function performs symmetric decryption based on the mode. */
+/* Error Returns Meaning */
+/* TPM_RC_FAILURE A fatal error */
+/* TPM_RCS_SIZE dSize is not a multiple of the block size for an algorithm that requires it */
+LIB_EXPORT TPM_RC
+CryptSymmetricDecrypt(
+		      BYTE                *dOut,          // OUT: decrypted data
+		      TPM_ALG_ID           algorithm,     // IN: the symmetric algorithm
+		      UINT16               keySizeInBits, // IN: key size in bits
+		      const BYTE          *key,           // IN: key buffer. The size of this buffer
+		      //     in bytes is (keySizeInBits + 7) / 8
+		      TPM2B_IV            *ivInOut,       // IN/OUT: IV for decryption.
+		      TPM_ALG_ID           mode,          // IN: Mode to use
+		      INT32                dSize,         // IN: data size (may need to be a
+		      //     multiple of the blockSize)
+		      const BYTE          *dIn            // IN: data buffer
+		      )
+{
+    INT16                blockSize;
+    BYTE                *iv;
+    BYTE                 defaultIv[MAX_SYM_BLOCK_SIZE] = {0};
+    evpfunc              evpfn;
+    EVP_CIPHER_CTX      *ctx = NULL;
+    int                  outlen1, outlen2;
+    BYTE                *pOut = dOut;
+    BYTE                *buffer = NULL;
+    UINT32               buffersize = 0;
+    BYTE                 keyToUse[MAX_SYM_KEY_BYTES];
+    UINT16               keyToUseLen = (UINT16)sizeof(keyToUse);
+    TPM_RC               retVal = TPM_RC_SUCCESS;
+
+    // These are used but the compiler can't tell because they are initialized
+    // in case statements and it can't tell if they are always initialized
+    // when needed, so... Comment these out if the compiler can tell or doesn't
+    // care that these are initialized before use.
+    pAssert(dOut != NULL && key != NULL && dIn != NULL);
+    if(dSize == 0)
+	return TPM_RC_SUCCESS;
```

Comment:
It is leagl to request decryption of a 0 size buffer ?

Reply:
This header part with all these tests is a copy of the original function.